src/artifacts_poetry_plugin/plugin.py

```diff
@@ -4,23 +4,141 @@
 import poetry.console.application
 from poetry.factory import Factory
 from cleo.commands.command import Command
+import poetry.config.config
+from poetry.publishing.uploader import Uploader
+from poetry.utils.authenticator import Authenticator
+from poetry.core.packages.package import Package
+from cleo.helpers import argument
+from typing import List
+import logging
 
-ARTIFACTS_LIST = "artifacts-list"
+ARTIFACTS_DEPLOY = "artifacts-deploy"
+REPOSITORY_NAME = "repository_name"
+logger = logging.getLogger(__name__)
 
 
-def getPackages(poetry):
+def get_repository_data(config, repo_name, io):
+    """Gets the URL, username, and password from the poetry configuration for the given repo name"""
+    repo_data = {}
+    repo_data["url"] = config.get(f"repositories.{repo_name}.url")
+    if repo_data["url"] is None:
+        raise RuntimeError(f"Repository {repo_name} is not defined")
+    authenticator = Authenticator(config, io)
+    auth = authenticator.get_http_auth(repo_name)
+    if auth:
+        repo_data["username"] = auth.username
+        repo_data["password"] = auth.password
+    else:
+        raise RuntimeError(
+            f"username and password are required. Please verify that your poetry config for this repository is correct for repository {repo_name}"
+        )
+    return repo_data
+
+
+def get_packages(poetry) -> List[Package]:
     locker = poetry.locker
     repository = locker.locked_repository()
     return repository.packages
 
 
+def get_cached_wheel_files(cache_dir):
+    """Gets all wheel file locations for the given cache directory"""
+    wheel_files = {}
+    for root, dirs, files in os.walk(cache_dir):
+        for file in files:
+            if file.endswith(".whl"):
+                wheel_files[file] = Path(os.path.join(root, file))
+    return wheel_files
+
+
+def add_package_paths(packages: List[Package], wheel_files):
+    """Adds the wheel file locations for each package's files"""
+    non_deployable_packages = []
+    deployable_packages = []
+    for package in packages:
+        file_found = False
+        for file in package.files:
+            if file["file"] in wheel_files:
+                file_found = True
+                file["url"] = wheel_files[file["file"]]
+        if file_found:
+            deployable_packages.append(package)
+        else:
+            non_deployable_packages.append(package)
+    return deployable_packages, non_deployable_packages
+
+
+def upload_packages(poetry, packages: List[Package], url, username, password, io):
+    uploader = Uploader(poetry=poetry, io=io)
+    uploader.auth(username=username, password=password)
+    session = uploader.make_session()
+    for package in packages:
+        uploader._package = package
```

Comment:
Q: Very strange that we're modifying the private properties of this object. Is there no other way to provide the package? Even if we had to create a new uploader for each package, that seems preferable to reaching into a private variable (unless it incurs some expensive cost.)